### PR TITLE
Improved logging

### PR DESCRIPTION
**Pull request by Skylake143**
This pull requests adds improved and cleaned up scripts as well as improved logging. During training both robust accuracy and clean accuracy are logged and displayed in wandb. Also in the end two more extensive metric tables are created containing training information such as training time, early stopping information and device info and final model performances on the train and validation sets. This improves usability of the tool vastly. Furthermore, the models were renamed and cleaned up in order to follow the same scheme. 
It was tested using the mnist_relu_2 scripts and on a CNN-4 architecture using the pgd script including arguments.

### CODE REVIEW

File: adversarial_training_box/models/CIFAR/densenet.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    def forward(self, x):
+        return self.down_sample(x)
+
+#DesneNet-BC
+#B stands for bottleneck layer(BN-RELU-CONV(1x1)-BN-RELU-CONV(3x3))
+#C stands for compression factor(0<=theta<=1)
+class DenseNet(nn.Module):
+    def __init__(self, block, nblocks, growth_rate=12, reduction=0.5, num_class=100, name = "densenet"):
+        super().__init__()
+        self.name = name
```
Comment (Aaron99B) [inline adversarial_training_box/models/CIFAR/densenet.py:69]: Why not a fixed name here like with the other networks?

File: adversarial_training_box/models/CIFAR/densenet.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        return output
+
+    def _make_dense_layers(self, block, in_channels, nblocks):
+        dense_block = nn.Sequential()
+        for index in range(nblocks):
+            dense_block.add_module('bottle_neck_layer_{}'.format(index), block(in_channels, self.growth_rate))
+            in_channels += self.growth_rate
+        return dense_block
+
+def densenet121(num_classes=10):
```
Comment (Aaron99B) [inline adversarial_training_box/models/CIFAR/densenet.py:120]: Or shall the name parameter be set here?

File: adversarial_training_box/models/CIFAR/cnn_small.py [hunk 3]
```diff
@@ -2,11 +2,12 @@
 import torch.nn as nn
 import torch.nn.functional as F
 
-class CNN(torch.nn.Module):
+class CNN_SMALL(torch.nn.Module):
     """Basic CNN architecture."""
 
     def __init__(self, in_channels=1):
-        super(CNN, self).__init__()
+        self.name = "cnn_small"
```
Comment (Aaron99B) [inline adversarial_training_box/models/CIFAR/cnn_small.py]: I like the idea with the name as a default parameter. But then let's make it consistent for all the networks

File: adversarial_training_box/models/CIFAR/wideresnet.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        out = self.layer3(out)
+        out = F.relu(self.bn1(out))
+        pool_size = 8 if self.input_size == 32 else 7
+        out = F.avg_pool2d(out, pool_size)
+        out = out.view(out.size(0), -1)
+        out = self.linear(out)
+
+        return out
+    
+def wideresnet_28_10(num_classes=10):
```
Comment (Aaron99B) [inline adversarial_training_box/models/CIFAR/wideresnet.py:92]: I think if we include these predefined configurations of the networks, we should set the name parameters accordingly

File: adversarial_training_box/models/MNIST/cnn_madry_medium.py [hunk 5]
```diff
@@ -4,11 +4,11 @@
 
 # Architecture from Madry et al. Towards Deep Learning Models Resistant to Adversarial Attacks
 
-class CNNMADRY(torch.nn.Module):
+class CNN_MADRY_MEDIUM(torch.nn.Module):
 
     def __init__(self):
-        self.name = "cnn_madry"
-        super(CNNMADRY, self).__init__()
+        self.name = "cnn_madry_medium"
```
Comment (Aaron99B) [inline adversarial_training_box/models/MNIST/cnn_madry_medium.py]: Then let's move it to a parameter here

File: adversarial_training_box/models/MNIST/cnn_small.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            hidden layer: 50
+            output layer: for MNIST 10 (digits 0 to 9)
+
+        Parameters:
+            input_height & input_width: int
+                The size of the image
+            num_classes: int
+                The number of classes we want to predict
+        """
+        super(CNN_SMALL, self).__init__()
```
Comment (Aaron99B) [inline adversarial_training_box/models/MNIST/cnn_small.py:19]: Here the name is missing

File: adversarial_training_box/models/MNIST/cnn_yang_big.py [hunk 7]
```diff
@@ -4,10 +4,10 @@
 
 # Architecture CNN2 from Yang et al. A Closer Look at Accuracy vs. Robustness
 
-class CNNYANG(nn.Module):
+class CNN_YANG_BIG(nn.Module):
     def __init__(self):
-        self.name = "cnn_yang"
-        super(CNNYANG, self).__init__()
+        self.name = "cnn_yang_big"
```
Comment (Aaron99B) [inline adversarial_training_box/models/MNIST/cnn_yang_big.py]: Move to parameter

File: adversarial_training_box/pipeline/pipeline.py [hunk 8]
```diff
@@ -19,52 +23,122 @@ def __init__(self, experiment_tracker: ExperimentTracker, training_parameters: A
     def save_model(self, network):
         self.experiment_tracker.save_model(network)
 
-    def train(self, train_loader: torch.utils.data.DataLoader, network: torch.nn.Module, training_stack: list[int, TrainingModule], validation_module: TestModule = None, in_training_validation_loader: torch.utils.data.DataLoader = None, early_stopper: EarlyStopper = None):
+    def train(self, train_loader: torch.utils.data.DataLoader, network: torch.nn.Module, training_stack: list[int, TrainingModule], validation_module: TestModule = None, validation_loader: torch.utils.data.DataLoader = None, early_stopper = None):
```
Comment (Aaron99B) [inline adversarial_training_box/pipeline/pipeline.py]: Why was the typing for the early stopper removed?

File: adversarial_training_box/pipeline/pipeline.py [hunk 9]
```diff
@@ -19,52 +23,122 @@ def __init__(self, experiment_tracker: ExperimentTracker, training_parameters: A
...
         for epochs, module in training_stack:
             for epoch in range(0, epochs):
-                network.train()
-                train_accuracy = module.train(train_loader, network, self.optimizer, self.experiment_tracker)
+                train_accuracy, robust_accuracy = module.train(train_loader, network, self.optimizer, self.experiment_tracker)
                 validation_accuracy = train_accuracy
 
                 if not self.experiment_tracker is None:
-                    self.experiment_tracker.log({"train_accuracy" : train_accuracy})
+                    self.experiment_tracker.log({"train_accuracy" : train_accuracy, "train_robust_accuracy" : robust_accuracy, "epoch": epoch +1})
```
Comment (Aaron99B) [inline adversarial_training_box/pipeline/pipeline.py]: Let's make the name for the train_robust_accuracy key consistent with the variable name

File: adversarial_training_box/pipeline/pipeline.py [hunk 10]
```diff
@@ -19,52 +23,122 @@ def __init__(self, experiment_tracker: ExperimentTracker, training_parameters: A
...
 
                 if not self.scheduler is None:
                     self.scheduler.step()
                 
                 if validation_module:
                     network.eval()
-                    _, _, validation_accuracy = validation_module.test(in_training_validation_loader, network)
+                    _, _, validation_accuracy, validation_robust_accuracy, valid_loss  = validation_module.test(validation_loader, network)
```
Comment (Aaron99B) [inline adversarial_training_box/pipeline/pipeline.py]: valid_loss -> validation_loss

File: adversarial_training_box/pipeline/pipeline.py [hunk 11]
```diff
@@ -19,52 +23,122 @@ def __init__(self, experiment_tracker: ExperimentTracker, training_parameters: A
...
+
+                                if self.experiment_tracker:
+                                    self.experiment_tracker.log({"smoothed_val_robust_accuracy": smoothed_val_robust_accuracy})
+
+                                if early_stopper.early_stop(smoothed_val_robust_accuracy):
+                                    early_stopping = True
+                                    early_stopping_epoch = epoch + 1
+                                    print(f"Early stopped at epoch: {early_stopping_epoch}")
+                                    break
+                        else: # Early stopping call based on validation accuracy
```
Comment (Aaron99B) [inline adversarial_training_box/pipeline/pipeline.py:102]: I would remove comments like this. If they are neccessary, the code should be more readable instead of a comment

File: adversarial_training_box/pipeline/pipeline.py [hunk 12]
```diff
@@ -19,52 +23,122 @@ def __init__(self, experiment_tracker: ExperimentTracker, training_parameters: A
...
         
         network.eval()
         for module in testing_stack:
-
+        
             print(f'testing for attack: {module.attack} and epsilon: {module.epsilon}')
 
-            attack, epsilon, accuracy = module.test(test_loader, network)
-            self.experiment_tracker.log_test_result({"epsilon" : epsilon, "attack" : str(attack), "accuracy" : accuracy})
+            attack, epsilon, test_accuracy, robust_accuracy, valid_loss = module.test(test_loader, network)
```
Comment (Aaron99B) [inline adversarial_training_box/pipeline/pipeline.py]: full names: valid_loss -> validation_loss

File: adversarial_training_box/pipeline/standard_test_module.py [hunk 13]
```diff
@@ -1,46 +1,63 @@
...
                 output = network(perturbed_data)
                 _, adv_pred = output.data.max(1, keepdim=True)
 
-                correct += adv_pred.eq(labels_for_correct_predictions.data.view_as(adv_pred)).sum().item()
-
-            else:
+                correct_adversarial += adv_pred.eq(labels_for_correct_predictions.data.view_as(adv_pred)).sum().item()
 
-                correct += pred.eq(target.data.view_as(pred)).sum().item()
+        # Calculate accuracies
```
Comment (Aaron99B) [inline adversarial_training_box/pipeline/standard_test_module.py:49]: Also comments like this and the other ones in the code I would remove. They make it harder to maintain the code. In general making the code readable >> adding a comment

**Review by Aaron99B — COMMENTED** [review]
I would make the names for the networks and how they are handled consistent. Further, I recommend to never use a comment unless absolutely necessary. Many of the comments can be removed by making the code more readable

File: adversarial_training_box/models/CIFAR/densenet.py (hunk 1 above)
Comment (Skylake143) [reply adversarial_training_box/models/CIFAR/densenet.py:69]: Set the name parameter as in your examples in the main class. Made it consistent in a new commit.

File: adversarial_training_box/models/CIFAR/cnn_small.py (hunk 3 above)
Comment (Skylake143) [reply adversarial_training_box/models/CIFAR/cnn_small.py]: Made this consistent to the rest

File: adversarial_training_box/models/CIFAR/wideresnet.py (hunk 4 above)
Comment (Skylake143) [reply adversarial_training_box/models/CIFAR/wideresnet.py:92]: Included naming for predefined configurations and made consistent for different networks

File: adversarial_training_box/pipeline/pipeline.py (hunk 8 above)
Comment (Skylake143) [reply adversarial_training_box/pipeline/pipeline.py]: Experimented with an alternative early stopper. Put the typing in again

File: adversarial_training_box/pipeline/pipeline.py (hunk 11 above)
Comment (Skylake143) [reply adversarial_training_box/pipeline/pipeline.py:102]: I don't agree with this. I believe in this case the comments clarify what the if else block is for without having to think about it any further.

**Review by Aaron99B — APPROVED** [review]
